google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/AbstractStatementParser.java

```diff
@@ -361,8 +392,58 @@ ClientSideStatement getClientSideStatement() {
   static final Set<String> dmlStatements = ImmutableSet.of("INSERT", "UPDATE", "DELETE");
   private final Set<ClientSideStatementImpl> statements;
 
+  /** The default maximum size of the statement cache in Mb. */
+  public static final int DEFAULT_MAX_STATEMENT_CACHE_SIZE_MB = 5;
+
+  private static int getMaxStatementCacheSize() {
+    String stringValue = System.getProperty("spanner.statement_cache_size_mb");
+    if (stringValue == null) {
+      return DEFAULT_MAX_STATEMENT_CACHE_SIZE_MB;
+    }
+    int value = 0;
+    try {
+      value = Integer.parseInt(stringValue);
+    } catch (NumberFormatException ignore) {
+    }
+    return Math.max(value, 0);
+  }
+
+  private static boolean isRecordStatementCacheStats() {
+    return "true"
+        .equalsIgnoreCase(System.getProperty("spanner.record_statement_cache_stats", "false"));
+  }
+
+  /**
+   * Cache for parsed statements. This prevents statements that are executed multiple times by the
+   * application to be parsed over and over again. The default maximum size is 5Mb.
+   */
+  private final Cache<String, ParsedStatement> statementCache;
+
   AbstractStatementParser(Set<ClientSideStatementImpl> statements) {
     this.statements = Collections.unmodifiableSet(statements);
+    int maxCacheSize = getMaxStatementCacheSize();
+    if (maxCacheSize > 0) {
+      CacheBuilder<String, ParsedStatement> cacheBuilder =
+          CacheBuilder.newBuilder()
+              // Set the max size to (approx) 5MB (by default).
+              .maximumWeight(maxCacheSize * 1024L * 1024L)
+              // We do length*2 because Java uses 2 bytes for each char.
+              .weigher(
+                  (Weigher<String, ParsedStatement>)
+                      (key, value) -> 2 * key.length() + 2 * value.sqlWithoutComments.length())
```

Comment:
Why are we only considering `value.sqlWithoutComments` as there are other attributes also stored as value? Is it because those have constant small size?

Reply:
Yes, exactly. The other properties are all small, and all have fixed size. The `statement` (so the original SQL statement) is also not stored as part of the value in the cache, as that is the key that we are using.